Draft Proposal for Diagnostics Client Library (a.k.a. "Runtime Client Library")

sywhang opened the pull request:
This is a draft proposal for the diagnostics client library - which we've been calling the "runtime client library" till now - (yes the rename is part of the proposal but I really want us to stop calling it the runtime client library :p). 

Some of the sample code is still left TODO, but I think this is at a point where we can openly start discussing about the design choices.

cc @noahfalk @josalem @mikem8361 @tommcdon

loic-sharma commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,296 @@
+
+## API Descriptions
+
+At a high level, the DiagnosticsClient library provides a `CommandHandler` class for each of the command specified as part of the diagnostics IPC protocol described here: https://github.com/dotnet/diagnostics/blob/master/documentation/design-docs/ipc-protocol.md#commands. For example, `EventPipeCommandHandler` handles all the `EventPipe` commands that are specified, such as `CollectTracing` and ` ...
```
The end of this paragraph seems to be missing

loic-sharma commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,296 @@
+    /// <summary>
+    /// A utility class that contain various helper methods to interact with the diagnostics IPC.
+    /// </summary>
+    public class DiagnosticsIPCHelper
```
Nit: should this be `DiagnosticsIpcHelper`? I believe the standard style for initialisms is upper cased first letter, lower cased subsequent letters. Example: `HttpClient`.

loic-sharma commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,296 @@
+    /// <summary>
+    /// A class for handling EventPipe commands.
+    /// </summary>
+    public class EventPipeCommandHandler
```
Consumers of this class will have a hard time unit testing. Consider adding an interface or making methods virtual.

sywhang replied:
Thanks for the comments @loic-sharma - I'll dig on this one for a little bit before making changes. I'm leaning towards adding an interface for now though.

loic-sharma replied:
From my understanding, default interface methods were added to mitigate this exact problem. See: https://github.com/dotnet/corefx/pull/41949/files

noahfalk replied:
My understanding is the default interface method needs to be implementable in terms of the other methods already on the interface. That doesn't seem feasible here?

loic-sharma commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,296 @@
+    int processId = Int32.Parse(args[0]);
+    
+    // Create an EventPipe command handler for the given process ID
+    EventPipeCommandHandler handler = new EventPipeCommandHandler(processId);
```
Can I create an EventPipe command handler for a process on a different machine? Or a process in a docker container?

sywhang replied:
Not at the moment since this isn't a supported scenario at the moment. We're in the process of figuring it out.

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,296 @@
+    /// <summary>
+    /// An enum for all the supported diagnostics IPC command
+    /// </summary>
+    public enum DiagnosticsCommand
```
You can give an enum an explicit type, i.e., in this case `public enum DiagnosticsCommand : byte`, which will ensure it matches the size of the enum in the IPC protocol.  This would also simplify the conversions from raw number to enum to string, etc.

Additionally, is this something we want an end user to be worrying about?  Will a library author be messing with these directly or will they be exclusively using the higher-level abstractions?  If the latter, we could probably make these internal.

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,296 @@
+        /// <returns>
+        /// IEnumerable of all the active process IDs.
+        /// </returns>
+        public static IEnumerable<int> GetActiveProcesses();
```
Rather than deal with the literal representation of the command codes, could this use the enum values instead?

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,296 @@
+    /// <summary>
+    /// A class for handling EventPipe commands.
+    /// </summary>
+    public class EventPipeCommandHandler
```
nit: I think `Controller` might make more sense in this context than `CommandHandler`.

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,296 @@
+    /// <summary>
+    /// A class for handling EventPipe commands.
+    /// </summary>
+    public class EventPipeCommandHandler
```
Would this library be the place for a higher-level abstraction of something like "session" for EventPipe?  For some of my GCDump work, I put together a crude session wrapper (see: https://github.com/josalem/diagnostics/blob/2e8e4f3576a45e51292bfd9c71cb80b272258d36/src/Tools/dotnet-trace/CommandLine/Commands/DotNetHeapDumper.cs#L422-L461), but I could see extrapolating that to have a builder class, be disposable, and abstract away some of the implementation details of the IPC protocol.  The "session" abstraction would be implemented via this Controller class.

sywhang replied:
We chatted offline during coffee but reposting as a comment to open it up for discussion w/ wider audience :p 

The way I imagined this library is a minimal set of APIs for library authors to interact with the runtime via the diagnostics IPC protocol. I think some kind of session management *could* be useful but it depends on where we draw the line for something that all library authors would rely on. As far as I can tell your EventPipeSession class in the code you wrote for GCDump work, you keep a reference to an EventPipeEventSource instance, making it act as a higher level abstraction that handles both EventPipe specific commands as well as some of the event handling logic with EventPipeEventSource. By that point, we're bringing in dependency on the entire TraceEvent library for just EventPipe that some users of this library may not even use (i.e. someone who might be using it as a thin shim for profiler attach, or some dump trigger based on counter data, etc.), and that is the part I don't like. 

That being said, we already do have a dependency on TraceEvent (even though the surface is very thin and could be removed with some trivial work), so I'm not closing on the possibility of adding more higher level abstractions/features to enable easier use of the API. But at the moment I am leaning towards keeping the APIs thin and minimal, and throw more features at it if needed.

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,296 @@
+        /// <summary>
+        /// Add a provider to the EventPipe command
+        /// </summary>
+        public void AddProvider(
```
I think this function would make more sense on a higher-level abstraction (see comment above about that).

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,296 @@
+
+        public string Name { get; }
+
+        public string FilterData { get; }
```
I think we should make this an `IEnumerable<KeyValuePair<string, string>>` since that more closely maps to what the `EventSource` APIs are expecting.  The IPC protocol takes it as a single string, but escaping the input with quotes and escaping the quotes has proved confusing for users of `dotnet-trace` to say the least.  If we change the inputs to to the library, we can make the escaping and quoting an implementation detail rather than a concern of the end user.

sywhang commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,296 @@
+We may create additional namespaces under Microsoft.Diagnostics.Client for command-specific classes that may be useful. i.e. `Microsoft.Diagnostics.Client.EventPipe`.
+
+### VersionInfo
+This is a helper class that encodes the minimum and maximum supported versions of the diagnostics IPC protocol. This helps the developer to programmatically opt in/out of certain features easily. Currently both min/max would be 1.
```
Actually, I thought about this again and I don't think there is as much of a use case here as I initially (and mistakenly) thought. I'll kill this one. Thanks for spotting that.

noahfalk commented:
cc @davidfowl @shirhatti

sywhang commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,298 @@
+## Intro
+The Diagnostics Client Library (currently named as "Runtime Client Library") is a managed library that can be used to interact with the .NET runtime via the diagnostics IPC protocol as documented in https://github.com/dotnet/diagnostics/blob/master/documentation/design-docs/ipc-protocol.md.
+
+The name "Diagnostics Client Library" comes from the fact that we call the runtime (CoreCLR) component responsible for accepting and handling the diagnostics IPC commands the "diagnostics server" - https://github.com/dotnet/coreclr/blob/master/src/vm/diagnosticserver.h. Since this library is a managed library on the other side of the IPC protocol responsible for communicating with the runtime's " ...
```
Microsoft.CoreCLR.DiagnosticsClient.dll sounds good to me as the assembly name. I’ll add that to the doc as well.

mikem8361 replied:
It isn't consist with the rest of our assemblies. They all start with Microsoft.Diagnostics... And having "CoreCLR" in the name seems like an internal name for .NET Core.  What about  Microsoft.Diagnostics.NETCore.Client?

sywhang replied:
How about Microsoft.NETCore.Diagnostics.Client?

josalem replied:
With the branding change of “.NET Core” -> “.NET” for 5.0, maybe we should avoid using “core” in the name.

sywhang replied:
The brand name might be changing to .NET but the diagnostics client only talks to CoreCLR and not any other .NET runtimes. (i.e. Mono) - unless supporting that is also a part of our plan..? I'm not aware of such plan though.

jorive replied:
IMO It's a .NET Core library, I do not see a reason to add the `NETCore` to the namespace.

sywhang replied:
I'm still inclined to put either NETCore or CoreCLR because just saying 'NET' seems too general and potentially misleading since this isn't compatible with other .NET runtimes.

noahfalk replied:
You should write the name of the assembly into the document : ) Its not always clear from the discussion what the plan of record is.

sywhang commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,298 @@
+
+## API Descriptions
+
+At a high level, the DiagnosticsClient library provides a `CommandHandler` class for each of the command specified as part of the diagnostics IPC protocol described here: https://github.com/dotnet/diagnostics/blob/master/documentation/design-docs/ipc-protocol.md#commands. For example, `EventPipeCommandHandler` handles all the `EventPipe` commands that are specified, such as `CollectTracing` and ` ...
```
I’ll go with ‘Client’ as you suggested :)

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,298 @@
+        public static IEnumerable<int> GetActiveProcesses();
+
+        // The default path that the diagnostics client library looks for the IPC socket (default: /tmp)
+        public static string DefaultIpcSocketPath { get; set; } 
```
Why is this static and mutable?

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,298 @@
+        /// <returns>
+        /// A Stream object to read
+        /// </returns> 
+        public Stream CollectTracing2()
```
What's the deal with these names?

sywhang replied:
They're the names given to the IPC protocol commands.. I'm not a fan of the names either but I think Noah suggested a good point so I can clean this up with better names.

josalem commented:
With the branding changing for .NET Core to just .NET for 5.0, could we choose a name that doesn’t have “core” in it?

________________________________
From: Sung Yoon Whang <notifications@github.com>
Sent: Tuesday, October 22, 2019 11:34:27 AM
To: dotnet/diagnostics <diagnostics@noreply.github.com>
Cc: John Salem <josalem@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [dotnet/diagnostics] Draft Proposal for Diagnostics Client Library (a.k.a. "Runtime Client Library") (#574)


@sywhang commented on this pull request.

________________________________

In documentation/design-docs/diagnostics-client-library.md<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fdiagnostics%2Fpull%2F574%23discussion_r337683050&data=02%7C01%7Cjosalem%40microsoft.com%7Ced062b946106451976da08d7571e7933%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637073660700082969&sdata=BVcg64Vr%2Bdj2mi7ivVeEH9LnPe4%2BQjQPvEYg70M6clg%3D&reserved=0>:

> @@ -0,0 +1,298 @@
+# Diagnostics Client Library API Design
+
+## Intro
+The Diagnostics Client Library (currently named as "Runtime Client Library") is a managed library that can be used to interact with the .NET runtime via the diagnostics IPC protocol as documented in https://github.com/dotnet/diagnostics/blob/master/documentation/design-docs/ipc-protocol.md.
+
+The name "Diagnostics Client Library" comes from the fact that we call the runtime (CoreCLR) component responsible for accepting and handling the diagnostics IPC commands the "diagnostics server" - https://github.com/dotnet/coreclr/blob/master/src/vm/diagnosticserver.h. Since this library is a managed library on the other side of the IPC protocol responsible for communicating with the runtime's "diagnostics server", calling this the "Diagnostics Client Library" made sense.


How about Microsoft.NETCore.Diagnostics.Client?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fdiagnostics%2Fpull%2F574%3Femail_source%3Dnotifications%26email_token%3DAE5VXHKUWEU26ANC2JTHLJDQP5BTHA5CNFSM4JCLTXIKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCI2F27Y%23discussion_r337683050&data=02%7C01%7Cjosalem%40microsoft.com%7Ced062b946106451976da08d7571e7933%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637073660700092963&sdata=%2BM5xRZMgU6ehJW2Xvq3P3k8p0xgcuF7Mat1%2F5NEr4eQ%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAE5VXHL5LQJ6BRFHIJQRH4LQP5BTHANCNFSM4JCLTXIA&data=02%7C01%7Cjosalem%40microsoft.com%7Ced062b946106451976da08d7571e7933%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637073660700092963&sdata=62615ipkYS7LFl0DhgaskqD%2FiSW5OyUGxfEnUR3I0E8%3D&reserved=0>.

sywhang commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,298 @@
+    {
+        public EventPipeProvider(
+            string name,
+            ulong keywords = ulong.MaxValue,
```
This is the current default

sywhang commented:
Thanks for the feedback everyone! I've changed the proposal quite a bit to address most of the concerns raised. Some of the main changes are:
- Added few more sample usage cases
- API surface is a lot more "flat" now, with just a single `DiagnosticsClient` class that has static methods for invoking various diagnostics IPC protocol commands. There is no more `CommandHandler` classes. 
- Changed namespace to `Microsoft.Diagnostics.NETCore.Client` 
- Removed sub-namespaces like `Microsoft.Diagnostics.Client.EventPipe`. 

When you find time, please take a look at the updated version and let me know what you think!

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,329 @@
+            int circularBufferMB,
+            bool rundownRequested,
+            Stream stream
+        )
```
nit: missing a `{`

sywhang replied:
This is just the signature and doesn't include body.

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,329 @@
+        public ReadOnlyList<EventPipeProvider> Providers { get ; }
+        public int CircularBufferMB { get; }
+        public bool RundownRequested { get; }
+        public Stream Stream { get; };
```
Do we want to put a `Stop` method on here that calls through to the `client`?

sywhang replied:
We could. In that case I'd rather remove the `DiagnosticsClient.StopTracing` since that would cause confusion.

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,329 @@
+    /// A utility class that contain various helper methods to interact with the diagnostics IPC.
+    /// </summary>
+    public class ProcessLocator
+    {
```
We currently only look in `/tmp` on Linux, do we want to expand this to allow for the user to specify a directory to look for sockets in?  I'm in thinking about the container scenarios where we mount the container's `/tmp` to the host's.  If we make the search directory arbitrary, we could mount `/tmp` to wherever and the tools would still work.

sywhang replied:
We can add that in later unless it's in a dire need?

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,329 @@
+    }
+    try
+    {
+        source.Process();
```
I don't think you stop the `counterSession`, so I don't think this call will ever return.

josalem commented:
Whoops!  Meant to hit comment, not approve, but I'm not sure how to undo that.

noahfalk commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,327 @@
+        /// <returns>
+        /// An EventPipeSession object representing the EventPipe session that just started.
+        /// </returns> 
+        public static EventPipeSession StartTracing(int processId, IEnumerable<EventPipeProvider> providers, int circularBufferMB=256, bool requestRundown=true)
```
I'd suggest making these instance methods and use the constructor to handle connection information (processId). There is a decent chance that in the future we'd need richer connection information such as pipe name and it would be unfortunate if we had to create overloads of every API to support the different possible addressing schemes.

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,327 @@
+    /// <summary>
+    /// A utility class that contain various helper methods to interact with the diagnostics IPC.
+    /// </summary>
+    public class ProcessLocator
```
Why is this a separate API?

noahfalk replied:
I don't think it has to be, but you should propose what alternative you like better : ) I was contemplating that it could be a static API on the DiagnosticsClient if all the existing methods became instance methods.

noahfalk commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,327 @@
+        /// <param name="processId">Target process' ID</param>
+        /// <param name="attachTimeout">Timeout in seconds</param>
+        /// <param name="attachTimeout">
+        public static int AttachProfiler(int processId, uint attachTimeout, Guid profilerGuid, string profilerPath, byte[] additionalData);
```
```suggestion
        public static int AttachProfiler(int processId, TimeSpan attachTimeout, Guid profilerGuid, string profilerPath, byte[] additionalData);
```

noahfalk commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,327 @@
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    // Generic wrapper for exceptions thrown by this library
+    public class DiagnosticsClientExcpetion : Exception {} 
```
```suggestion
    public class DiagnosticsClientException : Exception {} 
```

noahfalk commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,327 @@
+    public class ServerNotAvailableException : DiagnosticsClientException {} 
+
+    // When the runtime responded with an error
+    public class ServerThrownException : DiagnosticsClientException {} 
```
```suggestion
    public class ServerErrorException : DiagnosticsClientException {} 
```

noahfalk commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,327 @@
+    {
+        public EventPipeProvider(
+            string name,
+            long keywords = ulong.MaxValue,
```
You mentioned these were the 'current' default... did you mean current in RuntimeClient source or elsewhere? From the API design perspective I give RuntimeClient's current source low weight, TraceEvent medium weight, and EventListener the highest weight because stuff that goes into BCL usually has the most review. EventListener uses EventKeywords.None as the default keyword value when not specified and doesn't provide a default level. That sounds pretty reasonable to me.

sywhang replied:
I changed it to match EventListener.

noahfalk commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,327 @@
+            string name,
+            long keywords = ulong.MaxValue,
+            EventLevel eventLevel = EventLevel.Verbose,
+            string filterData = null)
```
To match EventListener
```suggestion
            IDictionary<string,string> arguments = null)
```

noahfalk commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,327 @@
+
+        public string Name { get; }
+
+        public IEnumerable<KeyValuePair<string, string>> FilterData { get; }
```
```suggestion
        public IDictionary<string, string>> Arguments { get; }
```

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,327 @@
+
+        public int ProcessId { get; }
+        public int SessionId { get; } 
+        public IReadOnlyList<EventPipeProvider> Providers { get ; }
```
I think we will want to add APIs for changing these values _during_ a session.  Would we not want to be able to read these values as well before we update them?  I would imagine a consumer would maybe want to do something like
```csharp
session.AddProvider(p1);
if (session.BufferSize < y)
    session.SetBufferSize(y);
Thread.Sleep(10_000);
session.RemoveProvider(p1);
session.AddProvider(p2); // Where p2 is less chatty than p1
session.SetBufferSize(z); // where z < y
```

For example, when collecting a gcdump, we have to create multiple sessions to achieve what one mutable session could.  I definitely see your point though.

noahfalk replied:
I'd try to err on the side of keeping APIs simple. If I can't point to a clear and valuable scenario that needs the API then I'd rather not add it. Correcting for too few APIs later is easy in comparison to trying to deprecate an API we regret adding.

I'd also be worried that these getters have more complexity than it may first appear - they represent a cache in a distributed system. If we assume the backing config data for a session in a future runtime version has an IPC protocol to mutate it then the mutations aren't forced to be transmitted by calls through this API wrapper. The cache could become unsynchronized because of out-of-band updates. Another challenge is that in the future we might want to add an IPC protocol command that reads the session state in addition to the one that updates it. If we did that we'd either need to convert the behavior of the cached getter API to become an uncached API that behaves a little differently, or introduce multiple similarly named APIs to handle both cached and uncached variations.

sywhang commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,327 @@
+    };
+
+    IEnumerable<EventPipeProvider> cpuProvider = new List<EventPipeProvider>() { 
+        new EventPipeProvider("Microsoft-Windows-DotNETRuntime", ClrTraceEventParser.Keywords.Default, EventLevel.Informational)
```
Yup - I added it

sywhang commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,327 @@
+        source.Process();
+        ShouldExit.WaitOne();
+    }
+    catch (Exception e)
```
Thanks - I filed an issue regarding this here: https://github.com/microsoft/perfview/issues/1033

davidfowl commented:
The lack of var in the sample code makes this feel like java.

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+
+public void PrintRuntimeGCEvents(int processId)
+{
+    List<EventPipeProvider> providers = new List<EventPipeProvider>() {
```
var

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+            (int)EventLevel.Informational, (long)ClrTraceEventParser.Keywords.GC)
+    };
+
+    DiagnosticsClient client = new DiagnosticsClient(processId);
```
var

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+    };
+
+    DiagnosticsClient client = new DiagnosticsClient(processId);
+    EventPipeSession session = client.StartEventPipeSession(
```
var 😄

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+        false, // RequestRundown
+        256 // CircularBufferMB
+    );
+    EventPipeEventSource source = new EventPipeEventSource(session.EventStream);
```
var

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+
+public void TraceProcessForDuration(int processId, int duration, string traceName)
+{
+    List<EventPipeProvider> cpuProviders = new List<EventPipeProvider>() {
```
var

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+
+    Task copyTask = Task.Run(async () =>
+    {
+        using(FileStream fs = new FileStream(traceName, FileMode.Create, FileAccess.Write))
```
nit: space between using and (

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+    {
+        using(FileStream fs = new FileStream(traceName, FileMode.Create, FileAccess.Write))
+        {
+            await traceSession.stream.CopyToAsync(fs);
```
Is stream really lowercase?

noahfalk replied:
No, I think we renamed it "EventStream" and the sample needs to be updated.

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+using System.IO;
+using System.Threading.Task;
+
+public void TraceProcessForDuration(int processId, int duration, string traceName)
```
Make duration a TimeSpan

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+        if (obj.EventName.Equals("EventCounters"))
+        {
+            // I know this part is ugly. But this is all TraceEvent.
+            IDictionary<string, object> payloadFields = (IDictionary<string, object>)(obj.GetPayloadValueByName("Payload"));
```
var

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+            EventLevel.Informational
+        ),
+    };
+    DiagnosticsClient client = new DiagnosticsClient(processId);
```
var x 3

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+    DiagnosticsClient client = new DiagnosticsClient(processId);
+    EventPipeSession session = client.StartEventPipeSession(providers);
+    EventPipeEventSource source = new EventPipeEventSource(session.EventStream)
+    source.Dynamic.All += (TraceEvent obj) => {
```
make the brace style consistent

davidfowl commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -0,0 +1,284 @@
+
+public void TriggerDumpOnCpuUsage(int processId, int threshold)
+{
+    List<EventPipeProvider> providers = new List<EventPipeProvider>() {
```
var

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -19,46 +19,63 @@ The goal of this library is as following:
+            source.Process();
+        }
+        // NOTE: This exception does not currently exist. It is something that needs to be added to TraceEvent.
+        catch (EventStreamException e)
```
Do we need to create a new exception or could we just catch the exception that _caused_ the stream reading error?  I don't want to hide the underlying issue by having the `EventSource` simply say "something went wrong".

sywhang replied:
By `EventSource` do you mean `EventPipeEventSource`?

josalem replied:
whoops!  Yeah, I do.

sywhang replied:
> could we just catch the exception that caused the stream reading error

This part is a little ambiguous. I think of this `EventStreamException` as an exception that gets thrown by EventPipeEventSource when there is an exception while reading the stream because either the stream was shut down by something. There could be other exception, like incorrect or corrupted payload, which can probably be encapsulated by a different exception class. 

>  I don't want to hide the underlying issue 
The sample is meant to capture the intention to add more exception classes to EventPipeEventSource to make it not throw general Exception to prevent exactly this kind of things from happening.

josalem commented on documentation/design-docs/diagnostics-client-library.md:
```diff
@@ -251,17 +342,14 @@ This is a class to represent an EventPipeSession. It is meant to be immutable an
+        ///<summary>
+        /// Stops the given session
+        ///</summary>
+        public void Stop();
```
Without a constructor or initialize method, how will `Stop()` know what session id to send?

sywhang replied:
The constructor is internal. _sessionId is a private member
